File: LeetCode/Java/Solution/src/ios/ac/cn/ImplementStrStr.java
package ios.ac.cn;

public class ImplementStrStr {
	// 注意对空串的处理
	// 注意起始点的范围是 [0, la-lb]，不是　[0,la-lb)
	public String strStr(String haystack, String needle) {
		for (int i = 0; i <= haystack.length() - needle.length(); i++) {
			int j;
			for (j = 0; j < needle.length(); j++) {
				if (haystack.charAt(i + j) != needle.charAt(j)) {
					break;
				}
			}
			if (j == needle.length()) {
				return haystack.substring(i);
			}
		}

		return null;
	}
}

File: LeetCode/Python/implement_strstr.py
#!/usr/bin/env python
#coding: utf-8

class Solution:
    # @param haystack, a string
    # @param needle, a string
    # @return a string or None
    def strStr(self, haystack, needle):
        if haystack == None or needle == None: return None

        m, n = len(haystack), len(needle)
        if m == n == 0: return ""

        i = 0
        while i < m:
            if m - i < n: return None
            j, k = 0, i
            while j < n and k < m:
                if haystack[k] != needle[j]: break
                k += 1
                j += 1
            if j == n: return haystack[i:]
            i += 1
        return None

if __name__ == '__main__':
    s = Solution()
    assert None == s.strStr(None, None)
    assert None == s.strStr("ab", None)
    assert None == s.strStr(None, "ab")
    assert None == s.strStr("abcd", "e")
    assert "bac" == s.strStr("abac", "b")
    assert "ab" == s.strStr("ab", "ab")
    assert "" == s.strStr("", "")
